File: backend/spring-boot/src/main/java/com/runhwani/runmate/service/CourseService.java
package com.runhwani.runmate.service;

import com.runhwani.runmate.dto.request.course.CourseRequest;
import com.runhwani.runmate.dto.response.course.CourseDetailResponse;
import com.runhwani.runmate.dto.response.course.CourseLikeResponse;
import com.runhwani.runmate.dto.response.course.CourseResponse;
import com.runhwani.runmate.exception.EntityNotFoundException;

import java.io.IOException;
import java.util.List;
import java.util.UUID;

public interface CourseService {
    // 1. 코스 생성
    UUID createCourse(CourseRequest request, UUID userId) throws IOException;
    // 2. 코스 삭제
    void deleteCourse(UUID courseId, UUID userId)
        throws EntityNotFoundException, IOException;
    // 3. 코스 검색
    List<CourseResponse> searchCourses(UUID userId, String keyword);
    // 4. 최근 코스 조회
    List<CourseResponse> getRecentCourses(UUID userId);
    // 5. 내가 등록한 코스 조회
    List<CourseResponse> getCoursesCreatedBy(UUID userId);
    // 6. 모든 코스 조회
    List<CourseResponse> getAllCourses(UUID userId);
    // 7. 코스 상세 조회
    CourseDetailResponse getCourseDetail(UUID userId, UUID courseId);
    // 8. 코스 좋아요 업데이트
    CourseLikeResponse updateCourseLike(UUID userId, UUID courseId);
}

File: RunMate/data/src/main/java/com/D107/runmate/data/remote/api/CourseService.kt
package com.D107.runmate.data.remote.api

import com.D107.runmate.data.remote.common.ApiResponse
import com.D107.runmate.data.remote.common.ServerResponse
import com.D107.runmate.data.remote.request.course.CreateCourseRequest
import com.D107.runmate.data.remote.response.course.CourseDetailResponse
import com.D107.runmate.data.remote.response.course.CourseIdResponse
import com.D107.runmate.data.remote.response.course.CourseItemResponse
import com.D107.runmate.data.remote.response.course.CourseLikeResponse
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.PATCH
import retrofit2.http.POST
import retrofit2.http.Part
import retrofit2.http.Path
import retrofit2.http.Query

interface CourseService {
    @POST("courses/create")
    suspend fun createCourse(@Body createCourseRequest: CreateCourseRequest): Response<ServerResponse<CourseIdResponse>>

    @PATCH("courses/{courseId}/like")
    suspend fun updateCourseLike(@Path("courseId") courseId: String): Response<ServerResponse<CourseLikeResponse>>

    @GET("courses/{courseId}")
    suspend fun getCourseDetail(@Path("courseId") courseId: String): Response<ServerResponse<CourseDetailResponse>>

    @DELETE("courses/{courseId}")
    suspend fun deleteCourse(@Path("courseId") courseId: String): Response<ServerResponse<Any?>>

    @GET("courses/search")
    suspend fun searchCourse(@Query("keyword") keyword: String): Response<ServerResponse<List<CourseItemResponse>>>

    @GET("courses/all")
    suspend fun getAllCourseList(): Response<ServerResponse<List<CourseItemResponse>>>

    @GET("courses/recent")
    suspend fun getRecentCourseList(): Response<ServerResponse<List<CourseItemResponse>>>

    @GET("courses/my")
    suspend fun getMyCourseList(): Response<ServerResponse<List<CourseItemResponse>>>


}